rep-2004.rst

```diff
@@ -0,0 +1,527 @@
+
+  * e.g. ``navigation2``, ``rosbag2``, etc.
+
+* Quality Level 3:
+
+  * tooling quality packages
+  * e.g. ``ros2cli``, ``rviz``, ``rqt``, etc.
+
+* Quality Level 4:
+
+  * demos, tutorials, and experiments
+  * e.g. research packages, ``demo_nodes_cpp``, ``examples_rclcpp_minimal_publisher``, etc.
+
+While each quality level will have different requirements, it's always possible to overachieve in certain requirements even if other requirements prevent a package from moving up to the next quality level.
+
+Quality Level 1
+^^^^^^^^^^^^^^^
+
+This category should be used for packages which are required for a reasonable ROS system in a production environment.
+That is to say that after you remove development tools, build tools, and introspection tools, these packages are still left over as requirements for a basic ROS system to run.
+However, that does not mean that packages that would not normally fit this description should never be called 'Level 1'.
+If there is a need for a particular package in a reasonable production scenario, then that package should be considered for this category as well.
+However, packages which we consider essential to getting a robot up and running quickly, but perhaps is a generic solution to the problem should probably not start out as 'Level 1' due to the high effort in getting a package to 'Level 1' and maintaining it there.
+
+For example, the packages which provide intra-process communication, inter-process communication, generated message runtime code, node lifecycle, etc. should probably all be considered for 'Level 1'.
+However, a package which provides pose estimation (like ``robot_pose_ekf``\ ) is a generic solution for something that most people need, but is often replaced with a domain specific solution in production, and therefore it should probably not start out as 'Level 1'.
+However, it may upgrade to it at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+*Note: bullets below that start with [ROS Core], will be the prescription for what we do in the core packages in order to meet the associated requirements*
+
+Requirements to be considered a 'Level 1' package:
+
+* Version Policy:
+
+  * Must have a version policy (e.g. ``semver``)
+  * Must be at a stable version (e.g. for ``semver`` that means have a version >= 1.0.0)
+  * Must have a strictly declared public API
+  * Must have a policy for API stability
+  * Must have a policy for ABI stability
+  * Must have a policy that keeps API and ABI stability within a released ROS Distribution
```

Comment:
I don't know what that would look like. What kind of analysis/tools are you thinking about?

Reply:
I'm assuming that @gavanderhoorn is referencing something like this tool: https://lvc.github.io/abi-compliance-checker/

It could be up to package maintainers to opt-in to using an API/ABI reporting tool, or there could be a common tool hooked to some part of the release process (like bloom).

Reply:
Just to follow up on this, I feel like this is part of "how" not "what", so it should go in the developer guide, since others might want to use different tools to achieve the same thing and this is something that is likely to change over time (new tool, different advice) and so putting it in something that can be more easily changed than a REP is desirable.